src/client/CodeEditor/getOrCreateEditor.tsx

```diff
@@ -454,6 +458,67 @@ export const getOrCreateEditor = async ({
     extensions.push(copilot({ aiCopilotEndpoint }));
   }
 
+  // Widget appears after instances of "todo" in code editor, allowing AI to implement todo tasks when clicked
+  // TODO make widget only appear in comments(?)
+  class ToDoWidget extends WidgetType {
+    constructor() { super() }
+
+    eq(other: ToDoWidget) { return false }
+
+    toDOM() {
+      const wrap = document.createElement("i");
+      wrap.style.display = "inline-flex";  // prevent block expansion
+      wrap.style.alignItems = "center";
+      wrap.style.justifyContent = "center";
+      wrap.className = "icon-button icon-button-dark";
+      const reactContainer = document.createElement("div");
+      wrap.appendChild(reactContainer);
+      const root = createRoot(reactContainer);
```

Comment:
Whoah this is a cool approach! I have not seen this before - a new React root within a CodeMirror extension. This might be a nice way to clean up some other extensions too, which do pretty raw DOM manipulation with the DOM API directly.